### ISSUE TITLE
[entropy_src] RFC: Programming Model for FW_OV_ENTROPY_INSERT

### ISSUE THREAD

**Issue by martin-lueker** [lowRISC/opentitan#10388]
**TL;DR:** We want to make a new class of recoverable error in entropy_src for unexpected entry or exit from the FW_OV mode.

**Overview**                               
                                                                          
In `entropy_src`, we need to determine how to properly handle entry and  exit envents when activating the FW_OV modes.   There no intention to support dynamic switching in or out of this mode.  Thus we need to make  a plan for handling unexpected transitions in or out of this mode.        
                                                                          
**Background**                                                            
                                                                          
In `entropy_src`, the FW_OV pathway is meant to provide a mechanism for overriding the functions of the entropy_src health checks and conditioning algorithms, in case there are any unforeseen statistical defects in the  external RNG, or there are future security considerations which not adequately handled by the silicon IP.                                     

**Use Case**

The FW_OV pathway has two main components, which are both enabled by the [FW_OV_CONTROL CSR](https://docs.opentitan.org/hw/ip/entropy_src/doc/index.html#Reg_fw_ov_control)
- The FW_OV_MODE field enables the FW_OV functions and provides a read path for the raw RNG data, enough queueing capacity for 2048 bits of RNG data.  This queue can be read out through the [FW_OV_RD_DATA](https://docs.opentitan.org/hw/ip/entropy_src/doc/index.html#Reg_fw_ov_rd_data) register.   This register can be disabled by OTP.
- FW_OV_ENTROPY_INSERT field initiates a complete bypass of the `entropy_src`, whereby the health tests and SHA3 conditioning mechanism can be replaced with a firmware implementation.  If enabled with FW_OV_MODE, then data written to [FW_OV_WR_DATA](https://docs.opentitan.org/hw/ip/entropy_src/doc/index.html#Reg_fw_ov_wr_data) will override the IP processing of the `entropy_src`.

If this mechanism is invoked, it is assumed that the `entropy_src` HW implementation is unsuitable.  Thus it is envisioned that if needed, this mode will be selected before the IP is enabled, and the IP will stay in that state indefinitely.  There is no requirement to support toggling the FW_OV_ENTROPY_INSERT mode on or off while the IP is enabled.

**Problem Statement**

As this feature injects partial seeds into the pipeline while the HW conditioner is active, toggling this switch while this IP active will have unpredictable results. _Though it is not an intended use case, the response of the IP under these unusual conditions must be specified and tested._

If we allow toggling of the FW_OV_ENTROPY_INSERT bit while the `entropy_src` is enabled, it will be very difficult to predict the output seeds with out entirely modelling the entire pipeline (with complete cycle-accurate delay modelling) in DV.  Thus, scoreboarding incidental switches between the two modes is not feasible.

Thus we see a need to explicitly the programming model for `entropy_src` to explicitly _forbid_ transitions in FW_OV_ENTROPY_INSERT in the programming model when the IP is active, such as the ones shown in the following waveform. Currently such transitions are not recommended, but there is nothing in the design to forbid them.

![image](https://user-images.githubusercontent.com/47870387/151263697-bcda1342-a3b8-4381-a578-6563bc0bfb04.png)

If we make this change to the programming model we need to select a method for enforcing this programming model in the IP.  There are a couple broad options.

**Option A: Write protection of  FW_OV_CONTROL**

We could render the FW_OV_CONTROL register read only when the DUT is enabled. This approach is also used by OTBN and `keymgr`.  `keymgr` has a `cfg_regwen` register which is asserted when the DUT is enabled, and `entropy_src` can follow that example.

**Option B: Make illegal writes to FW_OV_CONTROL a new class of recoverable alert.**

This option makes a lot of sense given that the FW_OV pathway should only be used if the firmware is certified to handle this operation. Generally speaking it is probably safe to assume that violations of the programming models for _any_ OpenTitan IP would only happen due to bugs or glitches.  We see two possibilities for how such glitches occur:

- Intentional (but poorly implemented) Violation: Software intends to use FW override, but implemented it badly.   This is bad for security because the use case for FW_OV presumes that there is something insecure about the HW implementation.  So this case merits a recoverable alert.
- Unintentional (SW bugs or glitches): Also bad for security, and also deserves a recoverable alert.

**Initial Recommendation**
Of the two options above, I would like to recommend Option B, that we updated the `entropy_src` IP to make forbidden transitions of the FW_OV_CONTROL register a new class of recoverable alert.

What do people think of this proposed change?

**Comment by martin-lueker** [comment]
Many thanks goes to @rswarbrick and @weicaiyang for their earlier feedback!

**Comment by martin-lueker** [comment]
For future documentation purposes here is the wavedrom JSON for the figure above:
```
{ signal: [
  {name:"rst_n", wave:"0.1...........|.01|......."},
  {name:"module_enable", wave:"0..1..........|0..|.1....."},
  {name:"fw_ov_entropy_insert", wave:"0........1....|0.1|.....0."},
  {name: "CSRNG output", wave: "0..4.3.2.x....|.0.|.4.2.x."}
],
head: {text: "Forbidden FW_OV transitions (regardless of OTP status)"},
foot: {text: "Note: \"module_enable\" and \"fw_ov_entropy_insert\" are actually MuBi4 values."+
        "The output to CSRNG is poorly defined if fw_ov_entropy_insert switches while enabled."}}
```

**Comment by tjaychen** [comment]
@martin-lueker could you clarify a bit under what circumstances the writes would be considered legal and illegal? 
It sounds like what you're saying is that once the hardware is enabled, we should not be allowed to toggle the settings, and once disabled, it would be fine to change it again?
This sounds a lot like the other entropy_src configurations to me.

A recoverable error sounds fine to me if that prefer that approach.  Just note that I don't think our regtool natively supports the behavior you're talking about and we'd need to some tweaking with the logic outside.  I assume the behavior you want is that when enabled, if someone tries to change the control, they are not able to and they get a recoverable alert right?

**Comment by martin-lueker** [comment]
Hi @tjaychen, that is actually a good point, and yes you have it all exactly correct.   Once enabled, the hardware should not be messed with, and we would like to enforce that.

Now that you mention it, you are also correct that the same argument applies to all the other configurations.  This came to mind though because writing a scoreboard for this case is particularly challenging.  This is a very good point.

Our original thought was that having an recov_alert to monitor just this one register would be easy and inexpensive.  However you are correct that it would require some logic outside the register block for either implementation option.

Looking more closely at [the solution employed in keymgr](https://github.com/lowRISC/opentitan/blob/06d3a28fd6c98a721dc9766268950049a50f67ed/hw/ip/keymgr/rtl/keymgr.sv#L316), it seems that is a very compact solution which nicely leverages the REGWEN construction in the register block.    That however still doesn't help our (mild) security issues if the firmware has a bug and accidentally turns on the FW_OV features after enabling, though hardware can't prevent software bugs.

@mwbranstad what do you think of the keymgr implementation linked above?  It seems very straightforward.

**Comment by tjaychen** [comment]
yeah i think a recoverable alert is fine too. It's just because we have no way in our regtool to "prevent" a write to a register under certain conditions, you will probably need to implement this an external register, absorbing the `qe` output so on and so forth.

**Comment by martin-lueker** [comment]
> yeah i think a recoverable alert is fine too. It's just because we have no way in our regtool to "prevent" a write to a register under certain conditions, you will probably need to implement this an external register, absorbing the `qe` output so on and so forth.

The `keymgr` implementation is pretty simple, but it doesn't lend it self to what we need.   In `keymgr` there is a regwen that is 100% hardware controllable.  We want one that is hardware controllable (and hardware reversable), but if software flips it, it is `rw1c`, and so it stays low.  `regtool` won't do that for sure.

**Comment by tjaychen** [comment]
o i think what you're looking for is doable via regtool also. The main thing the regtool can't do is the recov_alert.

**Comment by martin-lueker** [comment]
I'm now looking at both options.  (Really I want whatever secure solution is the easiest to implement) 
Will regtool allow you to control a register with multiple REGWENs?

If not I can think up solutions. 

For instance you could: 
1. Make a CFG_REGWEN that is SW ro and HW rw.
2. Configure the HJSON so that all the configuration registers are controlled by HW_REGWEN
3. Make a SW_CFG_REGWEN that is SW w1c (but HW ro)
4. In RTL, connect CFG_REGWEN.d to be `!enabled && sw_cfg_regwen.q`

The effect of this configuration would be that the configurations would become _permanently_ locked if SW_CFG_REGWEN has been cleared, and _reversibly_ locked if the device is enabled.

WDYT?

**Comment by martin-lueker** [comment]
Naturally the names are up to debate here (and yes, Mark, whatever we do I volunteer to smush the solution through CI)

**Comment by tjaychen** [comment]
o that sounds good. Yes yo'ure right, the regtool is only able to support regwen's that are software locked, hardware unlocked (or vice versa). It doesn't have the understanding that since it was software that issued a lock, it should never be unlocked.  So to accomplish the behavior you want (irreversible should software choose to lock, reversible if it's simply enable / disable), i think what you're suggesting makes sense.

**Comment by moidx** [comment]
In terms of FIPS enablement we have to run KAT tests on the conditioner. The sequence of events is as follows: 

- Stop `entropy_src` after completing boot mode.
- Configure FW OVR to execute KAT test using `FW_OV_WR_DATA` and `ENTROPY_DATA` registers. 
- Enable continuous mode of execution (FW OVR disabled).

The following test configures `entropy_src` in FW OVR mode: https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/tests/entropy_src_fw_ovr_test.c

Currently the test does not make an attempt to toggle the FW OVR mode at runtime. FW OVR enabling/disabling requires disabling and reconfiguring the `entropy_src`.

**Comment by martin-lueker** [comment]
Hi @moidx, under the current proposal that is all fine.  (also, thanks for pointing out that test!)
We just want to make sure that the IP is in a known (stable = disabled) state whenever we switch modes.

**Comment by martin-lueker** [comment]
> You may not need multiple regwen's though. You could have one regwen that is `rw0c`, and just maintain some extra state bits on the side based on enable.

That is an interesting point, but I personally don't know how to cleanly disable writes without another REGWEN.  We don't have access to the actual `we` lines on the registers themselves.  The `keymgr` regwen approach is nice and clean!

**Comment by tjaychen** [comment]
> > You may not need multiple regwen's though. You could have one regwen that is `rw0c`, and just maintain some extra state bits on the side based on enable.
> 
> That is an interesting point, but I personally don't know how to cleanly disable writes without another REGWEN. We don't have access to the actual `we` lines on the registers themselves. The `keymgr` regwen approach is nice and clean!

i might not have been thinking about it too clearly. I was thinking you could have one regwen, and implement most of the logic as to the regwen's behavior outside.  So then you could look at qe in combination with the enable to decide the final value.  So if you see a qe and q = 0, then you can transition the regwen to permanently 0.  If you've never seen qe and q = 0, then you could let the enable rising / fall edges control the regwen.  Something along those lines... I think I at one point had something like this in keymgr but didn't need it after some programmer model changes.

**Comment by mwbranstad** [comment]
Just sync'ed with @martin-lueker on this thread. I think the external REGWEN solution solves a lot of issues pretty easily. Will look at implementation details a bit more but do not expect any complications.

**Comment by mwbranstad** [comment]
This really seems like a good solution to me, and a proper use of REGWEN.
I am now wondering if we want to do this for EDN and CSRNG as well. It would require splitting out the MODULE_ENABLE bit unfortunately, but it would be consistent across blocks.

**Comment by tjaychen** [comment]
@mwbranstad if this seems like a good general solution to you, would you be open to making it a prim cell? :)
then long term either others can re-use it, or we can absorb it into regfile proper.

**Comment by mwbranstad** [comment]
I was planning on prototyping this to make sure it is as simple as i think it is. once done, let's discuss what that might look like.

**Comment by tjaychen** [comment]
> I was planning on prototyping this to make sure it is as simple as i think it is. once done, let's discuss what that might look like.

sounds good, thanks mark.

**Comment by mwbranstad** [comment]
Not much code to it:

  logic es_regwen_hw;

  assign hw2reg.regwen_hw.de = 1'b1;
  assign hw2reg.regwen_hw.d = es_regwen_hw;
  assign  es_regwen_hw = reg2hw.regwen_sw.q && !es_enable_pfe;

**Comment by tjaychen** [comment]
just to confirm Mark, does this mean there are separate regwen_sw and
regwen_hw fields?
Originally I was thinking you could just have one regwen field, of course,
that would require more tracking externally.

**Comment by martin-lueker** [comment]
Hi @tjaychen, looking at the [register topology for rw1c regs](https://docs.opentitan.org/doc/rm/register_tool/subreg_rw1c.svg) 
from the regtool documentation, if you only use one, you can't achieve HW reversibility.   We could of course create a new register type or REGWEN interaction model within regtool, but that would be much more complex in scope.

**Comment by mwbranstad** [comment]
To keep it simple and useful, we need two REGWENs. The third is already place (REGWEN_ME) which some use case might use to truly lock everything down. however I do not see OT using this option so far.

**Comment by tjaychen** [comment]
sounds good. I'm not against whatever is here.
I should just clarify that I was thinking you could use `hwext` mostly to do this. 

So like i mentioned before, if it's `hwext` you have the freedom to a lot of different kinds of detection. 
This is fine too however.

**Comment by msfschaffner** [comment]
This issue has been addressed as part of https://github.com/lowRISC/opentitan/pull/10520.
